chips/litex_vexriscv/src/plic.rs

```diff
@@ -0,0 +1,93 @@
+// TODO: This module may require renaming, as in VexRiscv the
+// interrupt controller probably isn't really called "PLIC"
+
+pub unsafe fn next_pending() -> Option<usize> {
+    let pending_interrupts = vexriscv_irq_raw::irq_pending();
+
+    // This is essentially an inefficient version of C's find first
+    // set (ffs()) function, giving the index of the least significant
+    // bit that is set
+    (0..=31).find(|test| pending_interrupts & (1 << test) != 0)
+}
+
+/// Disable all interrupts
+pub unsafe fn disable_interrupts() {
+    vexriscv_irq_raw::irq_setie(false);
+}
+
+/// Enable all interrupts
+pub unsafe fn enable_interrupts() {
+    vexriscv_irq_raw::irq_setie(true);
```

Comment:
This doesn't seem right. You shouldn't be changing the global interrupt enable (MIE) in the interrupt controller. This should instead enable or disable interrupts in the interrupt controller.

If they are always enabled then this should be a nop.

Reply:
Right, I see. The global external interrupt source should not be controlled in the interrupt controller itself, but rather in the chip crate. This then reduces the features of the interrupt controller to masking interrupts and checking which are pending, which is fine. Individual peripherals have their own event manager, which they have full control over.

Reply:
@alistair23 I've reworked the interrupt handling mechanisms, they should now closely match the `e310x` implementation. LiteX-VexRiscv does not have an `mtimer` by default, so I've thrown that out, but we can always introduce it again when necessary. If these changes look good to you, feel free to resolve this.